{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "# PatMatch initial demo and introduction\n",
    "\n",
    "Installation has been done!  \n",
    "(See [below](#About-setting-up-this-repository-and-using-this-interface) to learn more in regards to that.)\n",
    "\n",
    "PatMatch is ready to run!\n",
    "\n",
    "### Demo with included, prepared sequence files\n",
    "\n",
    "Demonstrate it works by clicking in the below cell and then hit `shift-enter` or press the `Run` button on the toolbar above. This uses author-supplied, prepared sequence files. You should see some results if running this from an active Jupyter notebook being served from [here](http://beta.mybinder.org/v2/gh/fomightez/patmatch-binder/master?filepath=index.ipynb) via the MyBinder.org service. (Click [here](http://beta.mybinder.org/v2/gh/fomightez/patmatch-binder/master?filepath=index.ipynb) to launch the active notebook if you happen to be reading this in a static environment.)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ../patmatch_1.2/\n",
    "perl patmatch.pl -c \"CAAAGGAA\" test/ATH1_cdna_test.prepared 1 ids"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make new cells to run and interact as if you were on the command line. You can also access a Bash terminal window by pressing the `Jupyter` symbol in the upper, left at the top of this page, clicking on `New` on the right abve the diretory listing, and selecting `Terminal` from the drop-down menu. Use the command line directly there in that window.\n",
    "\n",
    "Refer to this [this page](https://www.yeastgenome.org/nph-patmatch#examples) for 'Supported Pattern Syntax/Examples/Limits' information.\n",
    "\n",
    "If you'd like to learn about integrating PatMatch with Python and more about running PatMatch in this environment, I suggest continuing on to the next notebook [PatMatch with Python](PatMatch%20with%20Python%20basics.ipynb).\n",
    "\n",
    "If you don't need a command-line version of this program. There are a number of places where web-served versions of software are already set up to work with specific genomes:\n",
    "\n",
    "*Pattern Matching with PatMatch web-based tools*\n",
    "\n",
    "* [S. cerevisiae genomes](https://www.yeastgenome.org/nph-patmatch)\n",
    "* [Human genome](https://humancyc.org/patmatch.shtml?organism=HUMAN)\n",
    "* [Arabidopsis genome](http://www.arabidopsis.org/cgi-bin/patmatch/nph-patmatch.pl)\n",
    "* [S. lycopersicum](http://solcyc.solgenomics.net/patmatch.shtml?organism=LYCO)\n",
    "* [Aspergillus genomes](http://www.aspergillusgenome.org/cgi-bin/PATMATCH/nph-patmatch)\n",
    "* [Candida genome](http://www.candidagenome.org/cgi-bin/PATMATCH/nph-patmatch)\n",
    "* [E. coli genome](https://ecocyc.org/patmatch.shtml?organism=ECOLI)\n",
    "* [B. subtilis genome](https://bsubcyc.org/patmatch.shtml?organism=BSUB)\n",
    "* [Pseudomonas genome](http://www.pseudomonas.com:1555/patmatch.shtml?organism=PSEUDO)\n",
    "\n",
    "\n",
    "\n",
    "The PatMatch software is described in [this scientific article](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1160129/).\n",
    "\n",
    "Users should cite:\n",
    "\n",
    "PatMatch: a program for finding patterns in peptide and nucleotide sequences.  \n",
    "Yan T, Yoo D, Berardini TZ, Mueller LA, Weems DC, Weng S, Cherry JM, Rhee SY.  \n",
    "Nucleic Acids Res. 2005 Jul 1;33(Web Server issue):W262-6. [PMID: 15980466](https://www.ncbi.nlm.nih.gov/pubmed/15980466)\n",
    "\n",
    "The license for the software is available in the `patmatch_1.2` directory of actively running binders.\n",
    "\n",
    "\n",
    "\n",
    "-----\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "-----\n",
    "\n",
    "\n",
    "\n",
    "-----\n",
    "### Where to next?\n",
    "\n",
    "A useful README file is available at the site with the software; it is shown below. I suggest reading it and following the links it suggests to learn more about using PatMatch. Additionally,  refer to this [this page](https://www.yeastgenome.org/nph-patmatch#examples) for 'Supported Pattern Syntax/Examples/Limits' information. If you'd like to learn about integrating PatMatch with Python and more about running PatMatch in this environment, I suggest continuing on to the notebook [PatMatch with Python](PatMatch with Python.ipynb). If you are familiar with command line software, but are curious how this interface was set-up, see [below](#About-setting-up-this-repository-and-using-this-interface).  \n",
    "Below is the readme that came with the software. &#x1F53B;\n",
    "\n",
    "----\n",
    "\n",
    "----\n",
    "\n",
    "\n",
    "## Patmatch 1.2 README\n",
    "\n",
    "About\n",
    "----\n",
    "\n",
    "This directory contains the program used by PatMatch to do the actual pattern  \n",
    "matching.  The program is written in C and Perl.  It has been tested on Solaris  \n",
    "and Linux systems only.  The program can only be used from a terminal window.  \n",
    "\n",
    "Note: PatMatch was also tested on MacOSX by Dr. Rodolfo Chaparro. The  \n",
    "developer tools from the Apple website  \n",
    "(http://developer.apple.com/tools) must be downloaded in order to  \n",
    "compile Patmatch on MacOSX.  \n",
    "\n",
    "PatMatch 1.x provides a more powerful pattern matching than PatMatch in that  \n",
    "it allows patterns containing wildcards to support mismatch options, and it now  \n",
    "supports patterns containing both regular expressions and mismatch options.  \n",
    "\n",
    "This program uses a modified version of the NR-grep software for pattern  \n",
    "matching by G. Navarro.  The original version of NR-grep can be found at  \n",
    "http://www.dcc.uchile.cl/~gnavarro/software/  \n",
    "\n",
    "This program was developed by Thomas Yan and Danny Yoo.  \n",
    "\n",
    "For more information, please see the paper:  \n",
    "Yan, T, Yoo, D, Berardini, T, Mueller LA, Weems, D, Weng, S, Cherry JM,  \n",
    "Rhee, SY (2005) PatMatch: a program for finding patterns in peptide and  \n",
    "nucleotide sequences.  Nucleic Acids Research. 33 Web Server issue:W262-6.  \n",
    "Link:  \n",
    "http://nar.oxfordjournals.org/cgi/reprint/33/suppl_2/W262?ijkey=ZEc5zYlbTzvtJH1&keytype=ref\n",
    "\n",
    "\n",
    "Perl Modules Needed\n",
    "-------------------\n",
    "\n",
    "FindBin  \n",
    "Parse::RecDescent\n",
    "\n",
    "\n",
    "Installation\n",
    "------------\n",
    "\n",
    "In a terminal window, type:  \n",
    "gzip -d patmatch_1.2.tar.gz  \n",
    "\n",
    "To extract the decompressed tar file's contents, type:  \n",
    "tar -xvf patmatch_1.2.tar  \n",
    "\n",
    "This creates a new directory called patmatch_1.2.  cd into this directory and  \n",
    "compile the program by typing:  \n",
    "make\n",
    "\n",
    "\n",
    "Usage\n",
    "-----\n",
    "\n",
    "Preparing FASTA datasets for PatMatch.  This must be done before running  \n",
    "PatMatch:\n",
    "\n",
    "patmatch.pl accepts unjustified FASTA files as sequence files.  An  \n",
    "unjustified FASTA file is one where each sequence is in one line.  A FASTA  \n",
    "file can be unjustified using the unjustify_fasta.pl program.  This is   \n",
    "necessary for nrgrep_coords to search for matches.  \n",
    "Usage: perl unjustify_fasta.pl <file1> <file2> <file3>...  \n",
    "Example: perl unjustify_fasta.pl ATH1.pep ATH1.cds  \n",
    "         creates in the same directory ATH1.pep.prepared and ATH1.cds.prepared\n",
    "\n",
    "Run PatMatch with the command:  \n",
    "\n",
    "perl patmatch.pl [residue] [pattern] [sequence file] [number of mismatches] [mismatch types]\n",
    "\n",
    "Residue types:  \n",
    "\n",
    "-n for nucleotide  \n",
    "-c for complementary strand  \n",
    "-p for protein  \n",
    "\n",
    "Mismatch types can be:\n",
    "\n",
    "ids for insertions, deletions, and substitutions  \n",
    "i   for insertions only  \n",
    "id  for insertions and deletions  \n",
    "etc.  \n",
    "\n",
    "\n",
    "Pattern Syntax\n",
    "--------------\n",
    "\n",
    "This program uses the same pattern syntax as the version of Patmatch available  \n",
    "on TAIR's website, http://arabidopsis.org/cgi-bin/patmatch/nph-patmatch.pl  \n",
    "\n",
    "The website contains information about the supported pattern syntax as well as  \n",
    "some example patterns.  \n",
    "\n",
    "When running a pattern, if you get an \"Invalid pattern syntax.  Please check your pattern and try again.\" error,  \n",
    "and you are certain of the correctness of your pattern, try enclosing just the pattern in double quotes, such as:  \n",
    "\n",
    "perl patmatch/patmatch.pl -p \"L{3,5}X{5}DGZ\" patlibs/ATH1_pep 1 ids  \n",
    "\n",
    "for an example search on a peptide pattern with 1 insertion/deletion/substitution allowed.  \n",
    "\n",
    "Bugs & Problems\n",
    "---------------\n",
    "\n",
    "Please report problems and bugs to  \n",
    "curator@arabidopsis.org  \n",
    "\n",
    "\n",
    "\n",
    "Thomas Yan, July 1, 2005  \n",
    "updated by Sue Rhee, August 16, 2006  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---------\n",
    "<font color=\"#aaa\">(That text line above concludes the official PatMatch README file.)</font>&#x1F53A;\n",
    "\n",
    "------\n",
    "\n",
    "-----------\n",
    "\n",
    "\n",
    "## About setting up this repository and using this interface\n",
    "\n",
    "This particular notebook page runs with access directly to the Bash shell. You can type in commands in the cells as if you were on the command line.\n",
    "\n",
    "If you'd like a more full-featured Jupyter environment, proceed onto [the next notebook page](PatMatch%20with%20Python%20basics.ipynb) where the cells are connected to the interactive Python shell, or press the `Jupyter` symbol in the upper, left at the top of this page and then click on `New` on the right above the directory listing, and selecting a Python notebook from the drop-down menu. In that interactive Python-based notebook, you'll need to either preface your code to be directed to the command line with exclamation points. Alternatively, entire cells can be designated be run as Bash script with the `%%bash` magic tag as the first line of the cell. Learn about that and many of the magics commands [here](https://www.dataquest.io/blog/jupyter-notebook-tips-tricks-shortcuts/).\n",
    "In order to set up an environment with Perl and the required modules specified, the `environment.yml` file was edited. The software was copied from ftp://ftp.arabidopsis.org/home/tair/Software/Patmatch/ to the repository, and during the spinning up of the Binder instance, the process triggers via the `postBuild` file unpacking and compiling using `make` running in the directory generated in the course of the unpacking.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
